### PR TITLE
RSDK-5007 Interpolate should be a method on Frames

### PR DESCRIPTION
**Pull request by biotinker**
Inputs are floats whose meaning is known by the specific frame they are intended for.

This PR moves Interpolation to be something that each frame knows how to do for itself, rather than assuming that all inputs are able to be interpolated from one set to another.

### CODE REVIEW

File: referenceframe/frame.go [hunk 1]
```diff
@@ -391,6 +420,19 @@ func (rf *rotationalFrame) Transform(input []Input) (spatial.Pose, error) {
 	return spatial.NewPoseFromOrientation(&spatial.R4AA{input[0].Value, rf.rotAxis.X, rf.rotAxis.Y, rf.rotAxis.Z}), err
 }
 
+// Interpolate interpolates the given amount between the two sets of inputs.
+func (rf *rotationalFrame) Interpolate(from, to []Input, by float64) ([]Input, error) {
```
Comment (raybjork) [inline referenceframe/frame.go:424]: I think we should collapse the three functions in this file to a single one that operates on `baseFrame`

File: motionplan/rrtStarConnect.go [hunk 2]
```diff
@@ -119,7 +119,12 @@ func (mp *rrtStarConnectMotionPlanner) rrtBackgroundRunner(ctx context.Context,
 		}
 		rrt.maps = planSeed.maps
 	}
-	target := newConfigurationNode(referenceframe.InterpolateInputs(seed, rrt.maps.optNode.Q(), 0.5))
+	targetConf, err := mp.frame.Interpolate(seed, rrt.maps.optNode.Q(), 0.5)
+	target := newConfigurationNode(targetConf)
+	if err != nil {
```
Comment (nfranczak) [inline motionplan/rrtStarConnect.go:124]: [minor]: I think we can move this error check up one line

File: referenceframe/primitives.go [hunk 3]
```diff
@@ -64,10 +64,10 @@ type InputEnabled interface {
 	GoToInputs(context.Context, ...[]Input) error
 }
 
-// InterpolateInputs will return a set of inputs that are the specified percent between the two given sets of
+// interpolateAllInputs will return a set of inputs that are the specified percent between the two given sets of
```
Comment (raybjork) [inline referenceframe/primitives.go:67]: [opt] rename `interpolateInputs`? The "all" seems redundant

File: motionplan/solverFrame.go [hunk 4]
```diff
@@ -200,6 +200,32 @@ func (sf *solverFrame) Transform(inputs []frame.Input) (spatial.Pose, error) {
...
+func (sf *solverFrame) Interpolate(from, to []frame.Input, by float64) ([]frame.Input, error) {
+	if len(from) != len(sf.DoF()) {
+		return nil, frame.NewIncorrectInputLengthError(len(from), len(sf.DoF()))
+	}
+	if len(to) != len(sf.DoF()) {
+		return nil, frame.NewIncorrectInputLengthError(len(to), len(sf.DoF()))
+	}
+	interp := make([]frame.Input, 0, len(to))
+	posIdx := 0
+	for _, transform := range sf.frames {
```
Comment (nfranczak) [inline motionplan/solverFrame.go:213]: [q]: why do we call this "transform", I think we could use a slight better name here.

File: referenceframe/frame.go [hunk 5]
```diff
@@ -102,6 +102,9 @@ type Frame interface {
 	// Transform is the pose (rotation and translation) that goes FROM current frame TO parent's referenceframe.
 	Transform([]Input) (spatial.Pose, error)
 
+	// Interpolate interpolates the given amount between the two sets of inputs.
+	Interpolate(from, to []Input, by float64) ([]Input, error)
```
Comment (nfranczak) [inline referenceframe/frame.go:106]: major question: why is the output `[]Input` and not `[][]Input`? This seems like a mistake to me.
I assume we intend to replace 
```
interpolateSegment(ci *ik.Segment, resolution float64) ([][]frame.Input, error)
```
why is why I expect the outputs to match.

**Review by nfranczak — COMMENTED** [review]
Preliminary review. 
Will do a more in depth review once I understand why the new method on the frame interface is this
```
Interpolate(from, to []Input, by float64) ([]Input, error)
```
instead of this:
```
Interpolate(from, to []Input, by float64) ([][]Input, error)
```

**Review by raybjork — APPROVED** [review]
LGTM.  If you could look into the one comment I have about putting interpolate on the `baseFrame` instead I think that would be an improvement worth doing

File: motionplan/solverFrame.go (hunk 4 above)
Comment (raybjork) [reply motionplan/solverFrame.go:213]: agree.  Probably clearer to use `frame` so that we utilize the idiom of `for _, object := range objects`

Comment (biotinker) [reply motionplan/solverFrame.go:213]: Good call. This code dates back to early 2021 before we had a super strong concept of "frame" beyond just "spatial transform", now's a good time to update the nomenclature.

File: referenceframe/primitives.go (hunk 3 above)
Comment (biotinker) [reply referenceframe/primitives.go:67]: Yeah good call

File: referenceframe/frame.go (hunk 1 above)
Comment (biotinker) [reply referenceframe/frame.go:424]: I like that

File: referenceframe/frame.go (hunk 5 above)
Comment (biotinker) [reply referenceframe/frame.go:106]: Discussed in person

File: motionplan/rrtStarConnect.go (hunk 2 above)
Comment (biotinker) [reply motionplan/rrtStarConnect.go:124]: :+1:

**Review by nfranczak — APPROVED** [review]
🚀 LGTM
Just want to call out that can more an error check up one line in `rrtStarConnect.go`

**Pull request merged**